### PR TITLE
Callback listeners on put for all controllers

### LINKED ISSUE #236: Modify COAP and IP backends to send state right away like we did for BLE

### PR DESCRIPTION
**Pull request by bdraco**
closes #236

### CODE REVIEW

File: aiohomekit/controller/ip/pairing.py [hunk 1]
```diff
@@ -280,28 +275,33 @@ async def put_characteristics(self, characteristics):
...
-            iid = characteristic[1]
-            value = characteristic[2]
-            characteristics_set.add(f"{aid}.{iid}")
-            data.append({"aid": aid, "iid": iid, "value": value})
-        data = {"characteristics": data}
-
-        response = await self.connection.put_json("/characteristics", data)
+            aid, iid, value = characteristic
+            char_payload.append({"aid": aid, "iid": iid, "value": value})
+            listener_update[(aid, iid)] = {"value": value}
```
Comment (bdraco) [inline aiohomekit/controller/ip/pairing.py:283]: We should only add chars that we can read back

File: aiohomekit/controller/coap/pairing.py [hunk 2]
```diff
@@ -198,9 +190,22 @@ async def get_characteristics(
...
+    ) -> dict[tuple[int, int], dict[str, Any]]:
         await self._ensure_connected()
-        return await self.connection.write_characteristics(characteristics)
+        response_status = await self.connection.write_characteristics(characteristics)
+
+        listener_update = {tuple[int, int]: dict[str, Any]}
+        for characteristic in characteristics:
+            aid, iid, value = characteristic
+            if response_status.get((aid, iid), 0) == 0:
+                listener_update[(aid, iid)] = {"value": value}
```
Comment (bdraco) [inline aiohomekit/controller/coap/pairing.py:203]: We should only add chars that we can read back

**Comment by bdraco** [comment]
I just realized we are polling chars that send events as well.

We should probably only poll for chars that don't have the ev flag once we do the initial sync

**Comment by Jc2k** [comment]
> I just realized we are polling chars that send events as well.
> 
> 
> 
> We should probably only poll for chars that don't have the ev flag once we do the initial sync

I didn't do this originally until I had more confidence in the event code.

Now the biggest fear are stale TCP streams - polling has been an ok way of finding out if the other side crashes without doing an TCP RST on their side or if there was some other network failure (or they just got unplugged).

We could do a TCP keep alive but that's not part of spec so I don't know if it's widely support or any better than polling.

I think IP does have some version of the GSN, which made me wonder if it had zeroconf based disconnected events, but I'm not sure I saw it in the spec, and I'm not sure all my devices have the field in their zeroconf data.

**Comment by Jc2k** [comment]
Can't remember what we do atm, but for CoAP, it's UDP so no stream so I'm less worried about a stalled connection though still a bit worried about how we detect the device is unavailable

Would be better not to poll if we do because thread has battery powered devices.

**Comment by Jc2k** [comment]
Maybe the zeroconf recording dropping drives unavailability?

**Comment by bdraco** [comment]
> Maybe the zeroconf recording dropping drives unavailability?

That would make a lot of sense. If the zeroconf PTR expires we could ask the device if is still there and than mark it unavailable.

**Comment by bdraco** [comment]
Split the zeroconf / polling issue off into https://github.com/Jc2k/aiohomekit/issues/243

File: aiohomekit/controller/ip/pairing.py [hunk 3]
```diff
@@ -280,28 +276,36 @@ async def put_characteristics(self, characteristics):
...
-            iid = characteristic[1]
-            value = characteristic[2]
-            characteristics_set.add(f"{aid}.{iid}")
-            data.append({"aid": aid, "iid": iid, "value": value})
-        data = {"characteristics": data}
-
-        response = await self.connection.put_json("/characteristics", data)
+            aid, iid, value = characteristic
+            char_payload.append({"aid": aid, "iid": iid, "value": value})
+            accessory_chars = self.accessories.aid(aid).characteristics
```
Comment (bdraco) [inline aiohomekit/controller/ip/pairing.py:284]: This does a linear search under the hood which could be an issue with many acessories.  We should optimize that

Comment (bdraco) [reply aiohomekit/controller/ip/pairing.py:284]: https://github.com/Jc2k/aiohomekit/pull/248

**Comment by bdraco** [comment]
The tests that use the AccessoryServer kept failing so I went ahead and changed them to use the next available port instead and they seem to be a lot more reliable and faster now

**Comment by bdraco** [comment]
I've tested this with IP and BLE but I don't have COAP so I can't test there.

**Review by Jc2k — APPROVED** [review]

**Comment by Jc2k** [comment]
Working here with a CoAP Eve Thermo

**Pull request merged**